webui/react/src/hooks/useCustomizeColumnsModal.tsx

```diff
@@ -0,0 +1,234 @@
+import { ModalFunc } from 'antd/es/modal/confirm';
+import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import { FixedSizeList as List } from 'react-window';
+
+import { isEqual } from 'utils/data';
+import { camelCaseToSentence } from 'utils/string';
+
+import css from './useCustomizeColumnsModal.module.scss';
+
+interface ModalState {
+  columns: string[];
+  defaultVisibleColumns: string[];
+  initialVisibleColumns: string[];
+  onSave?: (columns: string[]) => void;
+  visible: boolean;
+}
+
+export interface ShowModalProps {
+  columns: string[];
+  defaultVisibleColumns: string[];
+  initialVisibleColumns: string[];
+  onSave?: (columns: string[]) => void;
+}
+
+interface ModalHooks {
+  showModal: (props: ShowModalProps) => void;
+}
+
+const useCustomizeColumnsModal = (): ModalHooks => {
+  const modalRef = useRef<ReturnType<ModalFunc>>();
+  const [ modalState, setModalState ] = useState<ModalState>({
+    columns: [],
+    defaultVisibleColumns: [],
+    initialVisibleColumns: [],
+    visible: false,
+  });
+  const [ searchTerm, setSearchTerm ] = useState('');
+  const [ visibleColumns, setVisibleColumns ] = useState<string[]>([]);
+
+  const showModal = useCallback((
+    { columns, initialVisibleColumns, defaultVisibleColumns, onSave }: ShowModalProps,
+  ) => {
+    setModalState({ columns, defaultVisibleColumns, initialVisibleColumns, onSave, visible: true });
+    setVisibleColumns(initialVisibleColumns);
+  }, []);
+
+  const closeModal = useCallback(() => {
+    if (!modalRef.current) return;
+    modalRef.current.destroy();
+    modalRef.current = undefined;
+  }, []);
+
+  const handleCancel = useCallback(() => {
+    if (!modalRef.current) return;
+    closeModal();
+  }, [ closeModal ]);
+
+  const handleSave = useCallback((state: ModalState) => {
+    if (!modalRef.current) return;
+    state.onSave?.(visibleColumns);
+    closeModal();
+  }, [ closeModal, visibleColumns ]);
+
+  const handleSearch = useCallback((e) => {
+    setSearchTerm(e.target.value.toLowerCase());
+  }, []);
+
+  const hiddenColumns = useMemo(() => {
+    const visibleColumnsSet = new Set(visibleColumns);
+    return modalState.columns.filter((column) => !visibleColumnsSet.has(column));
+  }, [ modalState.columns, visibleColumns ]);
+
+  const filteredHiddenColumns = useMemo(() => {
+    const regex = RegExp(searchTerm, 'i');
+    return hiddenColumns.filter(column => regex.test(camelCaseToSentence(column)));
+  }, [ hiddenColumns, searchTerm ]);
+
+  const filteredVisibleColumns = useMemo(() => {
+    const regex = RegExp(searchTerm, 'i');
+    return visibleColumns.filter(column => regex.test(camelCaseToSentence(column)));
+  }, [ visibleColumns, searchTerm ]);
+
+  const makeHidden = useCallback((transfer: string | string[]) => {
+    if (Array.isArray(transfer)) {
+      const transferSet = new Set(transfer);
+      setVisibleColumns(prev => prev.filter(column => !transferSet.has(column)));
```

Comment:
Nice! pretty nifty optimization, how did you come up with it?

Reply:
Searching stack overflow, haha! Plus remembering that Set.has() is an O(1) operation.